cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+        # body and use that instead.
+        elif isinstance(node.iter, ESingleton):
+            self.visit(subst(node.body, { node.id.id: node.iter.e }))
+            return
+
+        with self.stmt():
+            self.write("for ")
+            self.visit(node.id)
+            self.write(" in ")
+            self.visit(node.iter)
+            self.write(":")
+        with self.indented():
+            self.visit(node.body)
+
+    def visit_SAssign(self, node):
+        with self.stmt():
+            self.visit(node.lhs)
+            self.write(" = ")
+            self.visit(node.rhs)
+
+    def visit_SCall(self, node):
+        with self.stmt():
+            self.visit(node.target)
+            self.write(".", node.func)
+            with self.parens():
+                for arg in node.args:
+                    self.visit(arg)
+                    self.write(", ")
+
+    def visit_SMapDel(self, node):
+        with self.stmt():
+            self.write("del ")
+            self.visit(node.map)
+            self.write("[")
+            self.visit(node.key)
+            self.write("]")
+
+    def visit_SMapUpdate(self, node):
+        with self.stmt():
+            self.visit(node.val_var)
+            self.write(" = ")
+            self.visit(node.map)
+            self.write("[")
+            self.visit(node.key)
+            self.write("]")
+        self.visit(node.change)
+        # Writeback.
+        with self.stmt():
+            self.visit(node.map)
+            self.write("[")
+            self.visit(node.key)
+            self.write("] = ")
+            self.visit(node.val_var)
+
+    def visit_Op(self, node):
+        if node.assumptions:
+            import pdb; pdb.set_trace()
+        args = ", ".join([name for name, _ in node.args])
+        with self.python_block("def ", node.name, "(self, ", args, ")"):
+            # Since the body is a statement, we don't need to start the
+            # statement right here. Just go straight to the visitor. ~ C.
+            self.visit(node.body)
+            # And pass, in case nothing happened in the body.
+            self.write_stmt("pass")
+
+    def visit_Query(self, node):
+        if node.assumptions:
+            import pdb; pdb.set_trace()
+        args = ", ".join([name for name, _ in node.args])
+        with self.python_block("def ", node.name, "(self, ", args, ")"):
+            with self.stmt():
+                self.write("return ")
+                self.visit(node.ret)
+
+    def visit_Spec(self, spec, state_exps, sharing, abstract_state=()):
+        self.write_stmt("from rpython.rlib.objectmodel import specialize")
+
+        for func in spec.extern_funcs:
+            func_args = ", ".join([name for name, _ in func.args])
+            d = {n: n for n, _ in func.args}
+            with self.python_block("def ", func.name, "(", func_args, ")"):
+                # This is the style, so just go with it. ~ C.
+                body = func.body_string.format(**d)
```

Comment:
Yep, it was convenient to use Python's format strings as a way to define external functions. What you are doing here is right, even if it looks a little weird.

Reply:
Sweet! This arrangement should be good; it should give the RPython inliner full control over each external function.